Add Float16ArrayPixelBuffer

squelart opened the pull request:
#### 7857fde428154dcf9de6646518a5ae5f897df36e
<pre>
Add Float16ArrayPixelBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=283878">https://bugs.webkit.org/show_bug.cgi?id=283878</a>
<a href="https://rdar.apple.com/problem/140761117">rdar://problem/140761117</a>

Reviewed by Cameron McCormack.

Float16ArrayPixelBuffer is pretty much the same as
ByteArrayPixelBuffer, but obviously handling Float16Array data
instead.
Note: These types could eventually be reimplemented through
a single template, to be done later after they&apos;ve been used a
bit more...

At this time, Float16ArrayPixelBuffer is not actually used yet,
this will come in upcoming patches, with relevant tests.

Accompanying changes:
- Changed bool isByteArrayPixelBuffer() to a more flexible type()
  function, to be used by isType().
- Fixed safer-cpp warning in ByteArrayPixelBuffer::span().
- Made PixelFormat::RGBA16F officially supported in PixelBuffer.
- Updated PixelBuffer::computeBufferSize to handle more formats.
- Added PixelBuffer::computePixelComponentCount to distinguish
  which value (array size in bytes or color components) is needed
  in callers, since these two can now be different.
  Note: All callers have *not* been checked/updated yet! To be
  done in upcoming patches touching ImageData and 2d context...
  Related note for the future: It would be good to have strong
  types to help distinguish counts of pixels/components/bytes
  everywhere.
- Also added PixelBuffer::computePixelCount.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.cpp:
(WebCore::ByteArrayPixelBuffer::create):
(WebCore::ByteArrayPixelBuffer::tryCreate):
(WebCore::ByteArrayPixelBuffer::span const):
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.h:
(isType):
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp: Copied from Source/WebCore/platform/graphics/ByteArrayPixelBuffer.cpp.
(WebCore::Float16ArrayPixelBuffer::create):
(WebCore::Float16ArrayPixelBuffer::tryCreate):
(WebCore::Float16ArrayPixelBuffer::Float16ArrayPixelBuffer):
(WebCore::Float16ArrayPixelBuffer::createScratchPixelBuffer const):
(WebCore::Float16ArrayPixelBuffer::span const):
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.h: Copied from Source/WebCore/platform/graphics/ByteArrayPixelBuffer.h.
(WebCore::Float16ArrayPixelBuffer::data const):
(WebCore::Float16ArrayPixelBuffer::takeData):
(isType):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::PixelBuffer::supportedPixelFormat):
(WebCore::mustFitInInt32):
(WebCore::computeRawPixelCount):
(WebCore::computeRawPixelComponentCount):
(WebCore::PixelBuffer::computePixelCount):
(WebCore::PixelBuffer::computePixelComponentCount):
(WebCore::PixelBuffer::computeBufferSize):
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::type const):
(WebCore::PixelBuffer::isByteArrayPixelBuffer const): Deleted.

Canonical link: <a href="https://commits.webkit.org/287230@main">https://commits.webkit.org/287230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90fae450d9203e6af16f7a54765ad1f972eb5461

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57849 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83465 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30067 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6130 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61716 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19644 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81872 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71562 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42023 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25876 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70209 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26282 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84834 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6170 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4265 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69939 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69193 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17237 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13236 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11951 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6115 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12009 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6099 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9536 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/PixelBuffer.cpp:
```diff
@@ -38,30 +38,58 @@ bool PixelBuffer::supportedPixelFormat(PixelFormat pixelFormat)
+{
+    if (!uint32.hasOverflowed() && uint32.value() > uint32_t(std::numeric_limits<int32_t>::max()))
+        uint32.overflowed();
+    return uint32;
```
I see you are moving existing code, but you can use `isInBounds` from CheckedArithmetic.h to make this a little more concise:

```
if (!isInBounds<int32_t>(uint32.value()))
    uint32.overflowed();
return uint32;
```

or, and the compiler can probably optimize this to be equivalent to the above:

```
    return CheckedInt32(uint32);
```

heycam commented on Source/WebCore/platform/graphics/PixelBuffer.cpp:
```diff
@@ -38,30 +38,58 @@ bool PixelBuffer::supportedPixelFormat(PixelFormat pixelFormat)
 
+CheckedUint32 PixelBuffer::computeBufferSize(PixelFormat pixelFormat, const IntSize& size)
+{
+    unsigned bytesPerPixelComponent =
+#if HAVE(HDR_SUPPORT)
+        (pixelFormat == PixelFormat::RGBA16F) ? 2 :
+#endif
+        1;
```
I thought there might be an existing function around that does this, but I couldn't find one.

squelart replied:
Yeah, same. Ideally we'd want some table/function that knows the sizeof of each pixel format, but I decided not to, because:
- We'd need to work out the sizeof of all possible formats, even if some probably wouldn't be used anywhere(?).
- I'm not sure what it would be for (unsupported) RGB10 and RGB10A8. 😅 Or does it even make sense, if components are packed in bit ranges inside a bigger pixel bundle? Anyway, too much to worry about when it's not needed here.

I'll add a FIXME to revisit.

heycam commented on Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.h:
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2022 Apple Inc.  All rights reserved.
```
2024

heycam commented on Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp:
```diff
@@ -0,0 +1,117 @@
+
+std::optional<Ref<Float16ArrayPixelBuffer>> Float16ArrayPixelBuffer::create(const PixelBufferFormat& format, const IntSize& size, std::span<const Float16> data)
+{
+    // FIXME: Support non-8 bit formats.
```
This comment doesn't apply here, and I guess the one in ByteArrayPixelBuffer.cpp is now no longer needed.

heycam commented on Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp:
```diff
@@ -0,0 +1,117 @@
+
+RefPtr<Float16ArrayPixelBuffer> Float16ArrayPixelBuffer::tryCreate(const PixelBufferFormat& format, const IntSize& size)
+{
+    ASSERT(supportedPixelFormat(format.pixelFormat));
```
This assertion doesn't seem right, since we want to check if the pixel format is supported by `Float16ArrayPixelBuffer` specifically. And similarly now in `ByteArrayPixelBuffer`, which we don't want to support `PixelFormat::RGBA16`.

squelart replied:
I'd like to keep this debug-time check, to make sure that the expected format here is supported by the overall `PixelBuffer::supportedPixelFormat()`.

But indeed we could also strengthen the parameter check to make sure we are called with exactly RGBA16F here.
(And similarly in the old ByteArray one.)
I was initially considering removing this parameter since it looks useless here, but decided to keep it (at least for now), in case present/future generic code really wants to provide a format.

heycam commented on Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp:
```diff
@@ -0,0 +1,117 @@
+std::span<const uint8_t> Float16ArrayPixelBuffer::span() const
+{
+    Ref data = m_data;
+    ASSERT(data->byteLength() == (m_size.area() * 4 * sizeof(Float16)));
```
TIL we have a `Float16` type in the tree.

squelart committed:
Add Float16ArrayPixelBuffer
https://bugs.webkit.org/show_bug.cgi?id=283878
rdar://problem/140761117

Reviewed by Cameron McCormack.

Float16ArrayPixelBuffer is pretty much the same as
ByteArrayPixelBuffer, but obviously handling Float16Array data
instead.
Note: These types could eventually be reimplemented through
a single template, to be done later after they've been used a
bit more...

At this time, Float16ArrayPixelBuffer is not actually used yet,
this will come in upcoming patches, with relevant tests.

Accompanying changes:
- Changed bool isByteArrayPixelBuffer() to a more flexible type()
  function, to be used by isType().
- Fixed safer-cpp warning in ByteArrayPixelBuffer::span().
- Made PixelFormat::RGBA16F officially supported in PixelBuffer.
- Updated PixelBuffer::computeBufferSize to handle more formats.
- Added PixelBuffer::computePixelComponentCount to distinguish
  which value (array size in bytes or color components) is needed
  in callers, since these two can now be different.
  Note: All callers have *not* been checked/updated yet! To be
  done in upcoming patches touching ImageData and 2d context...
  Related note for the future: It would be good to have strong
  types to help distinguish counts of pixels/components/bytes
  everywhere.
- Also added PixelBuffer::computePixelCount.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.cpp:
(WebCore::ByteArrayPixelBuffer::create):
(WebCore::ByteArrayPixelBuffer::tryCreate):
(WebCore::ByteArrayPixelBuffer::span const):
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.h:
(isType):
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp: Copied from Source/WebCore/platform/graphics/ByteArrayPixelBuffer.cpp.
(WebCore::Float16ArrayPixelBuffer::create):
(WebCore::Float16ArrayPixelBuffer::tryCreate):
(WebCore::Float16ArrayPixelBuffer::Float16ArrayPixelBuffer):
(WebCore::Float16ArrayPixelBuffer::createScratchPixelBuffer const):
(WebCore::Float16ArrayPixelBuffer::span const):
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.h: Copied from Source/WebCore/platform/graphics/ByteArrayPixelBuffer.h.
(WebCore::Float16ArrayPixelBuffer::data const):
(WebCore::Float16ArrayPixelBuffer::takeData):
(isType):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::PixelBuffer::supportedPixelFormat):
(WebCore::mustFitInInt32):
(WebCore::computeRawPixelCount):
(WebCore::computeRawPixelComponentCount):
(WebCore::PixelBuffer::computePixelCount):
(WebCore::PixelBuffer::computePixelComponentCount):
(WebCore::PixelBuffer::computeBufferSize):
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::type const):
(WebCore::PixelBuffer::isByteArrayPixelBuffer const): Deleted.

Canonical link: https://commits.webkit.org/287230@main
